src/lib/scheduler/abstract_scheduler.cpp

```diff
@@ -0,0 +1,45 @@
+#include "abstract_scheduler.hpp"
+
+namespace opossum {
+
+void AbstractScheduler::wait_for_tasks(const std::vector<std::shared_ptr<AbstractTask>>& tasks) {
+  DebugAssert(([&]() {
+                for (const auto& task : tasks) {
+                  if (!task->is_scheduled()) {
+                    return false;
+                  }
+                }
+                return true;
+              }()),
+              "In order to wait for a task’s completion, it needs to have been scheduled first.");
+
+  // In case wait_for_tasks() is called from a Task being executed in a Worker, let the Worker handle the join()-ing,
+  // otherwise join right here
+  const auto worker = Worker::get_this_thread_worker();
+  if (worker) {
+    worker->_wait_for_tasks(tasks);
+  } else {
+    for (const auto& task : tasks) task->_join();
+  }
+}
+
+void AbstractScheduler::_group_tasks(const std::vector<std::shared_ptr<AbstractTask>>& tasks) const {
+  // Do nothing - grouping tasks is implementation-defined
```

Comment:
Why can't we make this pure virtual then?

Reply:
Because then you would force every scheduler to implement it. The ImmediateExecutionScheduler doesn't care about this.